File: app/src/main/java/cn/alvkeke/dropto/ui/adapter/CategoryListAdapter.java
package cn.alvkeke.dropto.ui.adapter;

import android.graphics.Color;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import cn.alvkeke.dropto.R;
import cn.alvkeke.dropto.data.Category;

public class CategoryListAdapter extends SelectableListAdapter<Category, CategoryListAdapter.ViewHolder> {

    public static class ViewHolder extends RecyclerView.ViewHolder {

        private final ImageView ivIcon;
        private final TextView tvTitle;
        private final TextView tvPreview;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            ivIcon = itemView.findViewById(R.id.item_category_icon);
            tvTitle = itemView.findViewById(R.id.item_category_title);
            tvPreview = itemView.findViewById(R.id.item_category_preview_text);

        }

        public void setTitle(String title) {
            tvTitle.setText(title);
        }

        public void setPreview(String preview) {
            tvPreview.setText(preview);
        }

        public void setType(Category.Type type) {
            ivIcon.setImageResource(Category.typeToIconResource(type));
        }

    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.rlist_item_category, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder viewHolder, int i) {
        Category c = this.get(i);
        if (c == null) {
            Log.e(this.toString(), "cannot get category at " + i);
            return;
        }

        viewHolder.setTitle(c.getTitle());
        viewHolder.setPreview(c.getPreviewText());
        viewHolder.setType(c.getType());
        if (isSelected(i)) {
            // TODO: use another color for selected item
            viewHolder.itemView.setBackgroundColor(Color.LTGRAY);
        } else {
            viewHolder.itemView.setBackgroundColor(Color.TRANSPARENT);
        }
    }

}

File: app/src/main/java/cn/alvkeke/dropto/ui/adapter/CategoryListAdapter.kt
package cn.alvkeke.dropto.ui.adapter

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import cn.alvkeke.dropto.R
import cn.alvkeke.dropto.data.Category
import cn.alvkeke.dropto.data.Category.Companion.typeToIconResource

class CategoryListAdapter : FilterableListAdapter<Category, CategoryListAdapter.ViewHolder>() {

    private var itemTouchHelper: ItemTouchHelper? = null
    fun setItemTouchHelper(helper: ItemTouchHelper) {
        this.itemTouchHelper = helper
    }
    var showDragHandler = false
        @SuppressLint("NotifyDataSetChanged")
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    val categories: List<Category>
        get() = elements

    val filteredCategory: List<Category>
        get() = filteredElements

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val ivIcon: ImageView = itemView.findViewById(R.id.item_category_icon)
        val tvTitle: TextView = itemView.findViewById(R.id.item_category_title)
        val tvPreview: TextView = itemView.findViewById(R.id.item_category_preview_text)
        val ivDragHandle: ImageView = itemView.findViewById(R.id.item_category_drag_handler)

        fun setTitle(title: String) {
            tvTitle.text = title
        }

        fun setPreview(preview: String) {
            tvPreview.text = preview
        }

        fun setType(type: Category.Type) {
            ivIcon.setImageResource(typeToIconResource(type))
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.rlist_item_category, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, i: Int) {
        val c = this.get(i)

        viewHolder.setTitle(c.title)
        viewHolder.setPreview(c.previewText)
        viewHolder.setType(c.type)
        if (showDragHandler) {
            viewHolder.ivDragHandle.visibility = View.VISIBLE
            viewHolder.ivDragHandle.setOnLongClickListener {
                itemTouchHelper?.startDrag(viewHolder)
                true
            }
        } else {
            viewHolder.ivDragHandle.visibility = View.GONE
        }
    }
}
